draft-ietf-mls-protocol.md

```diff
@@ -2323,25 +2314,52 @@ extensions can be updated, a GroupContextExtensions proposal is invalid if it
-
-A new group may be tied to an already existing group for the purpose of
-re-initializing the existing group, or to branch into a sub-group.
-Re-initializing an existing group may be used, for example, to restart the group
-with a different ciphersuite or protocol version. Branching may be used to
-bootstrap a new group consisting of a subset of current group members, based on
-the current group state.
-
-In both cases, the `psk_nonce` included in the `PreSharedKeyID` object must be a
-randomly sampled nonce of length `KDF.Nh` to avoid key re-use.
-
-### Sub-group Branching
-
-If a client wants to create a subgroup of an existing group, they MAY choose to
-include a `PreSharedKeyID` in the `GroupSecrets` object of the Welcome message choosing
-the `psktype` `branch`, the `group_id` of the group from which a subgroup is to
-be branched, as well as an epoch within the number of epochs for which a
-`resumption_secret` is kept.
+## Reinitialization
+
+A group may be reinitialized by creating a new group with the same membership
+and different parameters, and linking it to the old group via a resumption PSK.
+The members of a group reinitialize it using the following steps:
+
+* A member of the old group sends a ReInit proposal (see {{reinit}})
+* A member of the old group sends a Commit covering the ReInit proposal
+* The committer sends a Welcome message for the new group that matches the ReInit
+    * The `group_id`, `version`, and `cipher_suite` fields in the Welcome
+      message MUST be the same as the corresponding fields in the ReInit
+      proposal.
+    * The `epoch` in the Welcome message MUST be 1
+    * The Welcome MUST specify a PreSharedKey of type `resumption` with usage
+      `reinit`.  The `group_id` must match the old group, and the `epoch` must
+      indicate the epoch after the Commit covering the ReInit.
+    * The `psk_nonce` included in the `PreSharedKeyID` of the resumption PSK
+      MUST be a randomly sampled nonce of length `KDF.Nh`, for the KDF defined
+      by the new grou's ciphersuite.
+
+Resumption PSKs with usage `reinit` MUST NOT be used in other contexts.  A
+PreSharedKey proposal with type `resumption` and usage `reinit` MUST be
+considered invalid.
+
+## Sub-group Branching
+
+A new group can be formed from a subset of an existing group's members, using
+the same parameters as the old group.  The creator of the group indicates this
+situation by including a PreSharedKey of type `resumption` with usage `branch`
+in the Welcome message that creates the branched subgroup.
+
+A client receiving a Welcome including a PreSharedKey of type `resumption` with
+usage `branch` MUST verify that the new group reflects a subgroup branched from
+the referenced group.
+
+* The `version` and `ciphersuite` values in the Welcome MUST be the same as
+  those used by the old group.
+* Each KeyPackage in a leaf node of the new group's tree MUST be a leaf in the
+  old group's tree at the epoch indicated in the PreSharedKey.
```

Comment:
Do we really want the KeyPackages to be identical? This seems to constitute key-reuse across groups. This is where a (identity,endpoint_id) tuple would come in handy ;-) But I think going with the new multi-identity credentials, we could mandate here that the application has to ensure that the participants in the new group are a subset of the participants in the old group